src/fu-engine.c

```diff
@@ -2373,39 +2357,103 @@ fu_engine_sort_releases_cb (gconstpointer a, gconstpointer b)
-		g_autoptr(GError) error_local = NULL;
-		g_autoptr(FuInstallTask) task = fu_install_task_new (device, app_tmp);
-		if (!fu_engine_check_requirements (self, task,
-						   FWUPD_INSTALL_FLAG_ALLOW_REINSTALL |
-						   FWUPD_INSTALL_FLAG_ALLOW_OLDER,
-						   &error_local)) {
-			if (error_all == NULL) {
-				error_all = g_steal_pointer (&error_local);
-				continue;
-			}
-			/* assume the domain and code is the same */
-			g_prefix_error (&error_all, "%s, ", error_local->message);
+	if (!fu_engine_check_requirements (self, task,
+					   FWUPD_INSTALL_FLAG_ALLOW_REINSTALL |
+					   FWUPD_INSTALL_FLAG_ALLOW_OLDER,
+					   error))
+		return FALSE;
+
+	/* get all releases */
+	releases_tmp = xb_node_query (component,
+				      "releases/release",
+				      0, &error_local);
+	if (releases_tmp == NULL) {
+		if (g_error_matches (error_local, G_IO_ERROR, G_IO_ERROR_NOT_FOUND))
+			return TRUE;
+		if (g_error_matches (error_local, G_IO_ERROR, G_IO_ERROR_INVALID_ARGUMENT))
+			return TRUE;
+		g_propagate_error (error, g_steal_pointer (&error_local));
+		return FALSE;
+	}
+	for (guint i = 0; i < releases_tmp->len; i++) {
+		XbNode *release = g_ptr_array_index (releases_tmp, i);
+		GPtrArray *checksums;
+		g_autoptr(FwupdRelease) rel = fwupd_release_new ();
+
+		/* create new FwupdRelease for the XbNode */
+		fu_engine_set_release_from_appstream (self, rel, component, release);
+
+		/* invalid */
+		if (fwupd_release_get_uri (rel) == NULL)
 			continue;
+		checksums = fwupd_release_get_checksums (rel);
+		if (checksums->len == 0)
+			continue;
+
+		/* success */
+		g_ptr_array_add (releases, g_steal_pointer (&rel));
+	}
+
+	/* success */
+	return TRUE;
+}
+
+static gboolean
+fu_engine_add_releases_for_device_guid (FuEngine *self,
+					FuDevice *device,
+					const gchar *guid,
+					GPtrArray *releases,
+					GError **error)
+{
+	g_autofree gchar *xpath = NULL;
+	g_autoptr(GPtrArray) components = NULL;
+	g_autoptr(GError) error_local = NULL;
+
+	/* get all the components that provide this GUID */
+	xpath = g_strdup_printf ("components/component/"
+				 "provides/firmware[@type='flashed'][text()='%s']/"
+				 "../..", guid);
+	components = xb_silo_query (self->silo, xpath, 0, &error_local);
+	if (components == NULL) {
+		if (g_error_matches (error_local, G_IO_ERROR, G_IO_ERROR_NOT_FOUND)) {
+			g_set_error (error,
+				     FWUPD_ERROR,
+				     FWUPD_ERROR_NOTHING_TO_DO,
+				     "No releases found for device");
+			return FALSE;
 		}
-		return g_object_ref (app_tmp);
+		g_propagate_error (error, g_steal_pointer (&error_local));
+		return FALSE;
 	}
 
-	/* return the compound error */
-	g_propagate_error (error, g_steal_pointer (&error_all));
-	return NULL;
+	/* find all the components that pass all the requirements */
+	for (guint i = 0; i < components->len; i++) {
+		XbNode *component = XB_NODE (g_ptr_array_index (components, i));
+		if (!fu_engine_add_releases_for_device_component (self,
```

Comment:
This is adjusting the logic from previously isn't it?  Previously if a given requirement failed that component was skipped but the rest were processed.  Now you're aborting the whole method if the requirement failed on any one component.

Reply:
Yes, you're correct. I just wrote a long comment explaining the new logic, but in the process explained to myself that you're completely right! :) I'll redo the XPath query and make things a bit simpler.